### ISSUE TITLE
Remove the dependency on github.com/modern-go/reflect2

### ISSUE THREAD

**Issue by dominiquelefevre** [json-iterator/go#691]
`reflect2` is too brittle because of the reliance on the internals of the golang's compiler. Moreover, it is no longer maintained: [github.com/modern-go/reflect2.](https://github.com/modern-go/reflect2/issues/24#issuecomment-916998286).

Replace the usages of `github.com/modern-go/reflect2` with stdlib's `reflect`.

**Comment by fioncat** [comment]
Agreed, after i updated go to 1.20, `reflect2` panic directly, this is awful.

**Comment by pgimalac** [comment]
Also for the record `reflect2` wraps `reflect.Type` under the hood, which means when using it the linker considers the `Method` and `MethodByName` functions as reachable (even if this library doesn't use them), which disables some dead code elimination, resulting in a significantly bigger binary... (see [this talk](https://www.youtube.com/watch?v=EkG177eRcco) for more details)

Having this optimization enabled reduced a binary I work on from around 36MB to 29MB.
